### PR TITLE
Unify metrics behavior with dirs

### LINKED ISSUE #3930: dvc run -m/-M: inconsistent behavior with dirs
`dvc run` with `-M` flag doesn't accept directories as metrics whereas `dvc run` with `-m` flag does. This behavior should be unified.

**DVC version:**
- 0.94 (snap package)

**Additional Information :**
- OS: Ubuntu 18.04
- Discord context: https://discord.com/channels/485586884165107732/565699007037571084/717079938506686543
- Possibly related to #2973 

**Steps to reproduce**

1. Initialize git and dvc

2. Create `evaluate.py`:

```
import argparse
import os
import json

parser = argparse.ArgumentParser()
parser.add_argument("--metrics_dir", required=True)
args = parser.parse_args()
metrics_dir = args.metrics_dir

if not os.path.exists(metrics_dir):
    os.makedirs(metrics_dir)

with open(os.path.join(metrics_dir, "metrics1.json"), "w") as f:
    json.dump({"acc": 1.0}, f)

with open(os.path.join(metrics_dir, "metric2.json"), "w") as f:
    json.dump({"acc": 0.5}, f)
```

3. Create new stage with directory as cached metrics output:

```
$ dvc run -f metrics_cached.dvc -d evaluate.py -m metrics_cached  python3 evaluate.py --metrics_dir metrics_cached
Running command:
        python3 evaluate.py --metrics_dir metrics_cached

To track the changes with git, run:

        git add metrics_cached.dvc .gitignore
```

4. Create new stage with directory as not cached metrics output:

```
$ dvc run -f metrics_not_cached.dvc -d evaluate.py -M metrics_not_cached  python3 evaluate.py --metrics_dir metrics_not_cached
Running command:
        python3 evaluate.py --metrics_dir metrics_not_cached
ERROR: failed to run command - directory 'metrics_not_cached' cannot be used as metrics.
```

Both commands should either accept directories as metrics either discard them

### PR DESCRIPTION
**Pull request by nik123**
Fixes #3930

### CODE REVIEW

File: dvc/output/base.py [hunk 1]
```diff
@@ -256,6 +256,9 @@ def save(self):
                 )
             return
 
+        if self.metric:
+            self.verify_metric()
+
```
Comment (efiop) [inline dvc/output/base.py:261]: There is a
```
            if self.metric or self.plot:
                self.verify_metric()
```
a few lines above, under `if not self.use_cache`. How about we just move it outside of that if and check before that?

Comment (nik123) [reply dvc/output/base.py]: Fixed

**Comment by efiop** [comment]
For the record: partial fix for https://github.com/iterative/dvc/issues/3924

**Comment by efiop** [comment]
Looks great!

One more thing: could you please add a test for it?

**Comment by nik123** [comment]
@efiop , should I fix latest DeepSource issues too?

BTW it seems that PyTest preferred over the UnitTest and all the future tests should be implemented via PyTests. Is it correct?

**Comment by pmrowla** [comment]
> BTW it seems that PyTest preferred over the UnitTest and all the future tests should be implemented via PyTests. Is it correct?

Yes, this is correct

**Comment by efiop** [comment]
Thank you @nik123 ! :pray: 

> @efiop , should I fix latest DeepSource issues too?

It is advisory, but it is not required. I just did a quick-n-dirty conversion to pytest for now, no biggie :slightly_smiling_face:

**Pull request merged**